### ISSUE TITLE
Possible memory "leak" under Linux

### ISSUE THREAD

**Issue by gipfelen** [vapor/vapor#2493]
When returning a big amount of data within a single request call, memory will be allocated, but never gets released. The strange thing: It only happens on Linux and not on macOS.

I already tried to exchange postgres with SQLite, but that doesn't change anything.

### Steps to reproduce

Generate a small project with e.g.:

```swift
func getAllHandler(_ req: Request) throws -> EventLoopFuture<[Elaboration]> {        
    return Elaboration.query(on: req.db).all()
}
```

Which returns a lot of data (in my case more than 20.000 entries with several fields).

### Expected behavior

Release of memory after request is fulfilled, like in macOS. 

### Actual behavior

Memory will not get released

### Environment

Newest version of vapor 4.29.1, and all other dependencies with the newest version.

For building and launching vapor on Linux I used the latest docker container with the following tag: vapor/swift:5.2

Discussion in Discord: https://discordapp.com/channels/431917998102675485/519613337638797315/753882649780551731

**Comment by gipfelen** [comment]
I did some further digging and found out, that:
```
func getAllHandler(_ req: Request) throws -> EventLoopFuture<[Elaboration]> {        
	var elaborations: [Contact] = []

	for _ in 0...60000{
		let elaboration = Elaboration(name: "Test")
		
		elaborations.append(elaboration)
	}

	return req.eventLoop.makeSucceededFuture(elaborations)
}
```
also creates this leak. (more than 700mb not freed after this call)

**Comment by gipfelen** [comment]
I created a small test project to reproduce this issue: https://github.com/gipfelen/vaporLinuxLeak
simply call "docker-compose-up" and try to access http://localhost:8081/hello once and you should get a "leak" of 700MB

**Comment by rex-remind101** [comment]
Is there any plan to address this?

**Comment by fouadhatem** [comment]
I don't think this is limited to Linux only. How did you confirm that MacOS doesn't suffer from the same problem?

**Comment by gipfelen** [comment]
You can execute the program in macOS, and you will see, that the memory will be released after the api call.

**Comment by Joannis** [comment]
Was this compiled as DEBUG or RELEASE binary? I know that futures are held by SwiftNIO for a while as part of promise leak detection, which indirectly leads to appearant leaks.

**Comment by gipfelen** [comment]
It was the RELEASE binary.

**Comment by Joannis** [comment]
@gipfelen RELEASE as in the Swift compiler? Or RELEASE as in `swift build -c release`?

**Comment by fouadhatem** [comment]
Have you seen this: https://forums.swift.org/t/memory-leaking-in-vapor-app/22209?page=3

**Comment by gipfelen** [comment]
@Joannis RELEASE as in `swift build -c release`
@fouadhatem Yes, I saw it, but it has nothing to do with the database. Thus it has a different origin.

**Comment by Igor-Palaguta** [comment]
Have the same issue. After migrating from Vapor 3 to Vapor 4 can monitor constantly growing memory usage

**Comment by fouadhatem** [comment]
@siemensikkema @0xTim any thoughts on this?

**Comment by 0xTim** [comment]
Sorry for the delay. So I think this is the default behaviour for memory management on Linux in that the runtime won't release the memory unless the OS forces it to. @lukasa that's right isn't it?

**Comment by Lukasa** [comment]
The standard glibc `malloc` does not necessarily return memory to the OS when it is `free`d, yes. This is particularly true in cases where the heap is fragmented, in which case the memory _cannot_ be returned. glibc’s malloc does not particularly aggressively manage heap fragmentation so it’s a common source of concerns.

Both jemalloc and tcmalloc are known to be more aggressive in returning memory to the OS, so before conclusively diagnosing a memory leak on Linux I recommend running under one of those alternative allocators.

**Comment by Lukasa** [comment]
Additionally, while it will make program execution time much slower, Valgrind can be a useful tool for detecting memory leaks. Note that Swift does produce a couple of false positives under Valgrind, so care needs to be taken when analysing the results.

**Comment by obliqueRotationMethod** [comment]
I had a similar issue, memory use increased over time using Linux until the app crashed while the same code had no issues on macOS. I was able to solve this by implementing jemalloc in my container

Deployed my Vapor Docker Container on CloudFoundry with 1GB RAM allocated. Since last week I tested my code under production conditions, decoding an 15MB-ish JSON every hour and writing this data to an internal Fluent SQLite DB.  I noticed that now over time more and more RAM was used until the container crashed. 
@Lukasa mentioned jemalloc, hence I googled a bit and found that some other communities had similar issues and overcame them by using this allocator.
Using jemalloc my RAM is looking great, after 24hours at about 80MB used. Previously it would have crashed 3 times by now.

My Dockerfile:
[...]
# ================================
# Run image
# ================================
FROM swift:5.4-focal-slim

# Make sure all system packages are up to date.
RUN export DEBIAN_FRONTEND=noninteractive DEBCONF_NONINTERACTIVE_SEEN=true && \
    apt-get -q update && apt-get -q dist-upgrade -y **&& apt-get -y install libjemalloc2** && rm -r /var/lib/apt/lists/*

ENV LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libjemalloc.so**

[...]

**Comment by gipfelen** [comment]
I changed my  https://github.com/gipfelen/vaporLinuxLeak such that it uses jemalloc instead of malloc. But after a single call of http://localhost:8080/hello it uses more than 800MB of RAM, and it doesn't get freed (even after 5min +).

**Comment by Lukasa** [comment]
How much RAM is present on your system?

**Comment by Lukasa** [comment]
Also, is the application fully idle? If so, this is a known behaviour of jemalloc: jemalloc/jemalloc#1398

**Comment by obliqueRotationMethod** [comment]
@gipfelen Try changing 'libjemalloc2' to 'libjemalloc-dev' in the Dockerfile. This night I also saw that my App started crashing again when using libjemalloc2. Have rebuild it with libjemalloc-dev this morning and RAM is looking fine again

**Comment by gipfelen** [comment]
@Lukasa currently I'm testing it with Docker on macOS, where the VM is limited to 2GB of RAM. The CPU usage of the container is below 0.0% thus it also should be idling. 

@obliqueRotationMethod I tried to change it to 'libjemalloc-dev', but the behavior stays the same.

**Comment by Lukasa** [comment]
Right, so as linked above jemalloc will slowly _burn down_ freed memory over time, as the allocator is entered. A single request does not diagnose a leak. Try using the application consistently: I'd expect to see that long-term memory usage is flat.

**Comment by 0xTim** [comment]
Here's an excellent write up from @weissi about how the allocator works and how to see if it's actually leaking or not https://forums.swift.org/t/swift-rss-memory-consumption-is-much-greater-than-heap-memory/58837/10

**Comment by crarau** [comment]
The memory improvements are so significant!
Here is the before and after:
<img width="521" alt="image" src="https://user-images.githubusercontent.com/1430145/188981484-839b9738-392e-407d-ae92-4acbfe6606da.png">

I just did what @obliqueRotationMethod mentioned, added this in the Dockerfile

`
RUN export DEBIAN_FRONTEND=noninteractive DEBCONF_NONINTERACTIVE_SEEN=true && \
apt-get -q update && apt-get -q dist-upgrade -y && apt-get -y install libjemalloc-dev && rm -r /var/lib/apt/lists/*
ENV LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libjemalloc.so
`

Any reason why this is not included by default in the Vapor sample Dockerfile. 
We had to over-provision our infrastructure because of this issue costing us a lot of money.
It would be great if others don't have to go through the same pain.

**Comment by 0xTim** [comment]
The reason why it's not turned on by default in the Vapor template is it's not the default so is a bit of an unknown for us - I have no idea on the downsides of using libjemalloc instead of the default malloc. Any idea @Lukasa ?

**Comment by Lukasa** [comment]
jemalloc tends to trade off CPU for memory usage, so it tends to increase the baseline CPU usage a touch. Otherwise, there’s no real downside in almost all cases.

**Comment by crarau** [comment]
@0xTim @Lukasa in our tests we saw no noticeable impact on CPU and everything else is running smoothly (it's been almost 1 week now).

For reference I'm putting here our Dockerfile

```
# Dockerfile
ARG SWIFT_VERSION

# http://php.net/downloads.php
FROM swift:${SWIFT_VERSION}-focal

ARG env

ENV TZ=America/Vancouver
ENV DEBIAN_FRONTEND=noninteractive

RUN apt-get -qq update && apt-get install -y \
    apt-utils \
    libxml2 \
    libxml2-dev \
    pkg-config \
    libbsd0 \
    libcurl4-openssl-dev \
    libcurl4 \
    zlib1g-dev \
    libatomic1 \
    tzdata \
    uuid-dev \
    && rm -r /var/lib/apt/lists/* && apt-get clean

RUN export DEBIAN_FRONTEND=noninteractive DEBCONF_NONINTERACTIVE_SEEN=true && \
apt-get -q update && apt-get -q dist-upgrade -y && apt-get -y install libjemalloc-dev && rm -r /var/lib/apt/lists/*
ENV LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libjemalloc.so

WORKDIR /app
```

**Comment by obliqueRotationMethod** [comment]
@crarau nice to hear that this helped you

@0xTim @Lukasa I run this config in production since last November. Using Vapor as a backend to a jobApplication service for an international corporation. I implemented jemalloc at the beginning of my tests on the productive environment, faced no Vapor-related issues. Our nightly DB update now chews through multiple 20MB JSONs, thanks to jemalloc 1GB RAM is still plenty for the job. 
W/o changing the allocator we probably would have needed to increase the RAM already at launch because it crashed when parsing 3 fat JSONs in a row. 
Could be that CPU has a little more work. Aside from the parsing, CPU is idling at 3 to 5% load. So, I couldn't care less :)

**Comment by Lukasa** [comment]
My gut instinct is that jemalloc is likely to be a good default for Vapor applications: in general in containerized environments it is easier to trade RAM for CPU than the other way around.

**Comment by H0nestDude** [comment]
Thx a lot @obliqueRotationMethod ! Using jemalloc also saved me from memory leaks on linux while using postgres db ! ❤️

**Comment by fouadhatem** [comment]
@Lukasa and @0xTim , does this mean that jemalloc needs to become the default for Vapor on Linux moving forward?

**Comment by mkll** [comment]
An interesting observation. After jemalloc being injected in the Vapor process, each restart attempt end up with the following in the log:

```
Sep 17 20:20:16 node00 systemd[1]: Stopping SmartLoad backend daemon...
Sep 17 20:20:16 node00 smartload[16962]: [ NOTICE ] Server is shutting down...
Sep 17 20:20:16 node00 smartload[16962]: Received signal 11. Backtrace:
Sep 17 20:20:16 node00 smartload[16962]: 0x56200ced0c02, Backtrace.(printBacktrace in _B82A8C0ED7C904841114FDF244F9E58E)(signal: Swift.Int32) -> () at /home/admin/downloaded/smartload/.build/checkouts/swift-backtrace/Sources/Backtrace/Backtrace.swift:66
Sep 17 20:20:16 node00 smartload[16962]: 0x56200ced0ec7, closure #1 (Swift.Int32) -> () in static Backtrace.Backtrace.install(signals: Swift.Array<Swift.Int32>) -> () at /home/admin/downloaded/smartload/.build/checkouts/swift-backtrace/Sources/Backtrace/Backtrace.swift:80
Sep 17 20:20:16 node00 smartload[16962]: 0x56200ced0ec7, @objc closure #1 (Swift.Int32) -> () in static Backtrace.Backtrace.install(signals: Swift.Array<Swift.Int32>) -> () at /home/admin/downloaded/smartload/<compiler-generated>:79
Sep 17 20:20:16 node00 smartload[16962]: 0x7fcde789213f
Sep 17 20:20:16 node00 smartload[16962]: 0x7fcde7dcf258
Sep 17 20:20:16 node00 smartload[16962]: 0x56200dc45066, Vapor.Application.Sessions.Storage.__deallocating_deinit at /home/admin/downloaded/smartload/<compiler-generated>:0
Sep 17 20:20:16 node00 smartload[16962]: 0x7fcde7dcf25a
Sep 17 20:20:16 node00 smartload[16962]: 0x56200dc81301, destroy value witness for Vapor.Storage.Value at /home/admin/downloaded/smartload/<compiler-generated>:0
Sep 17 20:20:16 node00 smartload[16962]: 0x7fcde7dc5e0a
Sep 17 20:20:16 node00 smartload[16962]: 0x7fcde7baaad0
Sep 17 20:20:16 node00 systemd[1]: smartload.service: Main process exited, code=killed, status=11/SEGV
Sep 17 20:20:16 node00 systemd[1]: smartload.service: Failed with result 'signal'.
Sep 17 20:20:16 node00 systemd[1]: Stopped SmartLoad backend daemon.
```

**Comment by gohanlon** [comment]
I haven't tried `jemalloc` with Swift/Vapor yet, but I will experiment with adopting it in my own Vapor app soon.

In the past, adopting `jemalloc` on Linux has solved problems for me without creating any new ones. I also haven't personally observed a compromise with adopting `jemalloc`: I've generally gotten both higher performance and lower memory usage; and performance and memory usage have been more even over time. That's surely application specific, and my `jemalloc` experience isn't even with Swift.

@crarau's memory usage improvements are very interesting, and harmonizes with my (non-Swift) experience with `jemalloc`. If that improvement holds generally, Vapor adopting `jemalloc` by default could be compelling. And as @Lukasa mentioned, even more compelling in today's containerized envs/Firecracker VMs. 

Adopting `jemalloc` would be a large enough change that there would probably need to be a plan, e.g.:

* Alpha: Have several Vapor users to experiment with adopting `jemalloc` in their own apps and report anything interesting to this thread.
* Beta: Assuming things are going well, add a commented-out `jemalloc` section to the default `Dockerfile`, making it easy for users to "opt-in". Then, encourage users via the blog/Twitter to give `jemalloc` a spin.
* After enough information is collected, move forward:
  1. Make `jemalloc` the default.
  2. Recommend `jemalloc` as an opt-in, and note whatever caveats have been discovered.
  3. Remove `jemalloc` entirely.

**Comment by Lukasa** [comment]
@mkll Can you install debug symbols for jemalloc and try to catch that in a debugger? I’m curious what the actual issue is.

**Comment by mkll** [comment]
@Lukasa I'd love to, but I don't have much experience developing and debugging on Linux, and unfortunately don't have enough time to learn it right now. Maybe later, sorry.

**Comment by maciejtrybilo** [comment]
For what it's worth that's last 14 days of a service that had been causing me grief. I was restarting it after getting a memory usage alert.

It's running on Digital Ocean Apps on a 1GB RAM instance.

No prizes for spotting when we switched to `jemalloc` :)

<img width="1215" alt="Screenshot 2023-01-24 at 10 58 19" src="https://user-images.githubusercontent.com/1718852/214262374-2bdd58e1-9267-4c42-8374-63cb762cd472.png">

**Comment by lemonlab** [comment]
I experienced the same issue : memory leak **on Linux only**. Everything was OK on MacOS (monitoring with Instruments).

The app is hosted on [Digital Ocean App Platform](https://www.digitalocean.com/products/app-platform) on a Docker Debian instance.

![CleanShot 2023-11-22 at 09 14 31@2x](https://github.com/vapor/vapor/assets/2278002/5d34f8d2-70bd-4259-8723-42d104ee983f)

After using `jemalloc` to run the Docker image, everything went fine, i had no more memory leak.
I used `jemalloc` only to run the instance. For the build part, i've kept the standard Debian library.
